### PR TITLE
Mark unstable attribute and argument as deprecated

### PR DESCRIPTION
**Pull request by arezaii**
This PR adds deprecation warnings when using attributes with an argument that is not wrapped in parentheses. Previously we supported `@unstable "msg"`, but we want to drop that and move to supporting `@unstable("msg")` instead.

In https://github.com/chapel-lang/chapel/pull/21425 I removed support for `@unstable "msg"` without deprecating it. Since our release notes for 1.27/1.28 highlight this as available for users, it should be deprecated with a nice warning message.

While there, I also fixed some formatting and removed a previously satisfied `TODO`.

TESTING:

- [x] paratest w/futures `[Summary: #Successes = 14681 | #Failures = 0 | #Futures = 900]`
- [x] gasnet paratest (multilocale-only) `[Summary: #Successes = 1263 | #Failures = 0 | #Futures = 36]`
- [x] mac builds
- [x] linux builds

reviewed by @dlongnecke-cray - thank you!

### CODE REVIEW

File: frontend/include/chpl/parsing/parser-error-classes-list.h [hunk 1]
```diff
@@ -69,4 +69,5 @@ POSTPARSE_ERROR_CLASS(MultipleManagementStrategies, const uast::New::Management,
 POSTPARSE_ERROR_CLASS(PostParseErr, std::string)
 POSTPARSE_WARNING_CLASS(PostParseWarn, std::string)
 ERROR_CLASS(UnsupportedAsIdent, const uast::As*, const uast::AstNode*)
+WARNING_CLASS(ParenlessAttributeArgDeprecated, const uast::Attribute*)
```
Comment (DanilaFe) [inline frontend/include/chpl/parsing/parser-error-classes-list.h:72]: We try to keep the error classes lists sorted, both in the header and source files.

File: frontend/include/chpl/uast/Attribute.h [hunk 2]
```diff
@@ -33,27 +33,33 @@ class Attribute final: public AstNode {
   // the attribute name - deprecated or unstable or chpldoc.nodoc, for example
   UniqueString name_;
   int numActuals_; // number of child actuals
+  bool usedParens_; // whether the attribute was written with parens - only needed
+                    // while @unstable "msg" is going through deprecation from 1.30
   std::vector<UniqueString> actualNames_;
 
-  Attribute(UniqueString name, int numActuals, AstList actuals, std::vector<UniqueString> actualNames)
+  Attribute(UniqueString name, int numActuals, bool usedParens, AstList actuals,
```
Comment (dlongnecke-cray) [inline frontend/include/chpl/uast/Attribute.h:40]: Nit: I'd put used parens before `numActuals` so it lines up with the fields (and also generally with the actual token order).

File: frontend/include/chpl/uast/Attribute.h [hunk 3]
```diff
@@ -33,27 +33,33 @@ class Attribute final: public AstNode {
...
       numActuals_(numActuals),
+      usedParens_(usedParens),
       actualNames_(std::move(actualNames)) {
   }
 
   Attribute(Deserializer& des)
     : AstNode(asttags::Attribute, des) {
       name_ = des.read<UniqueString>();
       numActuals_ = des.read<int>();
+      usedParens_ = des.read<bool>();
```
Comment (dlongnecke-cray) [inline frontend/include/chpl/uast/Attribute.h:53]: Makes me wonder if we should encapsulate our fields in a macro to make this generally easier.

File: frontend/include/chpl/uast/Attribute.h [hunk 4]
```diff
@@ -161,11 +168,20 @@ class Attribute final: public AstNode {
...
     ser.write(numActuals_);
+    ser.write(usedParens_);
     ser.write(actualNames_);
   }
 
   DECLARE_STATIC_DESERIALIZE(Attribute);
 
+  /*
+    Returns whether the attribute was written with parens.
+    This is only needed while @unstable "msg" is going through deprecation
```
Comment (dlongnecke-cray) [inline frontend/include/chpl/uast/Attribute.h:179]: I'd drop the part of the comment mentioning 1.30-1.31, we can just silently remove the whole method when we're ready, so no need to date ourselves.

File: frontend/include/chpl/uast/Attribute.h [hunk 5]
```diff
@@ -33,27 +33,33 @@ class Attribute final: public AstNode {
   // the attribute name - deprecated or unstable or chpldoc.nodoc, for example
   UniqueString name_;
   int numActuals_; // number of child actuals
+  bool usedParens_; // whether the attribute was written with parens - only needed
```
Comment (dlongnecke-cray) [inline frontend/include/chpl/uast/Attribute.h:36]: Remove part mentioning we need it for 1.30, no need to date ourselves.

File: frontend/lib/parsing/ParserContext.h [hunk 6]
```diff
@@ -131,11 +131,14 @@ struct ParserContext {
...
 
   void noteAttribute(YYLTYPE loc, AstNode* firstIdent,
-                                  ParserExprList* toolspace,
-                                   MaybeNamedActualList* actuals);
+                     bool usedParns,
+                     ParserExprList* toolspace,
+                     MaybeNamedActualList* actuals);
+
   owned<Attribute> buildAttribute(YYLTYPE loc, AstNode* firstIdent,
+                                  bool usedParns,
```
Comment (dlongnecke-cray) [inline frontend/lib/parsing/ParserContext.h:139]: typo: usedParns

File: frontend/lib/parsing/ParserContext.h [hunk 7]
```diff
@@ -131,11 +131,14 @@ struct ParserContext {
   owned<AstNode> consumeVarDeclLinkageName(void);
 
   void noteAttribute(YYLTYPE loc, AstNode* firstIdent,
-                                  ParserExprList* toolspace,
-                                   MaybeNamedActualList* actuals);
+                     bool usedParns,
```
Comment (dlongnecke-cray) [inline frontend/lib/parsing/ParserContext.h:134]: typo: usedParns

File: frontend/lib/parsing/chpl.ypp [hunk 8]
```diff
@@ -878,7 +878,25 @@ attribute_decl_stmt_ls:
 attribute_decl_stmt:
  attribute_decl_begin opt_attribute_decl_toolspace opt_attribute_actuals
  {
-  context->noteAttribute(@1, $1, $2, $3);
+  context->noteAttribute(@1, $1, true, $2, $3);
+ }
+ | attribute_decl_begin opt_attribute_decl_toolspace
+ {
+  // this allows us to distinguish between @attribute() and @attribute
+  // but should not be needed once `@unstable "msg"` goes through its deprecation
```
Comment (dlongnecke-cray) [inline frontend/lib/parsing/chpl.ypp:886]: No need to date ourselves in a comment. You can make an issue to track this.

File: frontend/lib/parsing/chpl.ypp [hunk 9]
```diff
@@ -878,7 +878,25 @@ attribute_decl_stmt_ls:
...
+  // cycle from 1.30 - 1.31
+  auto lst = new MaybeNamedActualList();
+  context->noteAttribute(@1, $1, false, $2, lst);
+ }
+ | attribute_decl_begin STRINGLITERAL
+ {
+  // this allows us to maintain @unstable "message" without requiring parens
+  // but should not be needed once `@unstable "msg"` goes through its deprecation
```
Comment (dlongnecke-cray) [inline frontend/lib/parsing/chpl.ypp:894]: Ditto the 1.30~.

File: frontend/lib/parsing/parser-error-classes-list.cpp [hunk 10]
```diff
@@ -410,6 +410,14 @@ void ErrorMultipleManagementStrategies::write(ErrorWriterBase& wr) const {
   }
 }
 
+void ErrorParenlessAttributeArgDeprecated::write(ErrorWriterBase& wr) const {
```
Comment (dlongnecke-cray) [inline frontend/lib/parsing/parser-error-classes-list.cpp:413]: Nice.

File: frontend/test/parsing/testParseAttributes.cpp [hunk 11]
```diff
@@ -905,7 +905,33 @@ static void test18(Parser* parser) {
...
+    @deprecated "this thing is deprecated"
+    var x: int;
+    @stable "1.28"
+    var y: int;
+  )"""";
 
+  auto parseResult = parseStringAndReportErrors(parser, "test19.chpl",
+                                                program.c_str());
+  assert(guard.numErrors() == 3);
+  assert(guard.error(0)->message() == "Attribute arguments without parentheses are deprecated; please wrap the argument in parentheses '()'");
```
Comment (dlongnecke-cray) [inline frontend/test/parsing/testParseAttributes.cpp:927]: Nit: 80 cols (but should I even bother, lol)

**Review by dlongnecke-cray — APPROVED** [review]
Only comment is philosophical: In general, we should avoid timestamping/dating ourselves in code. Don't mention release dates and things as those get outdated/unsync'd really fast. Let the git history do that work for us (and make issues as needed).

**Pull request merged**